resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/Client.scala

```diff
@@ -107,33 +110,50 @@ private[spark] class Client(
 
     val k8ClientConfig = k8ConfBuilder.build
     Utils.tryWithResource(new DefaultKubernetesClient(k8ClientConfig)) { kubernetesClient =>
-      ShutdownHookManager.addShutdownHook(() =>
-        kubernetesResourceCleaner.deleteAllRegisteredResourcesFromKubernetes(kubernetesClient))
-      val sslConfigurationProvider = new SslConfigurationProvider(
-        sparkConf, kubernetesAppId, kubernetesClient, kubernetesResourceCleaner)
-      val submitServerSecret = kubernetesClient.secrets().createNew()
-        .withNewMetadata()
-          .withName(secretName)
-          .endMetadata()
-        .withData(Map((SUBMISSION_APP_SECRET_NAME, secretBase64String)).asJava)
-        .withType("Opaque")
-        .done()
-      kubernetesResourceCleaner.registerOrUpdateResource(submitServerSecret)
-      try {
-        val sslConfiguration = sslConfigurationProvider.getSslConfiguration()
-        // start outer watch for status logging of driver pod
-        val driverPodCompletedLatch = new CountDownLatch(1)
-        // only enable interval logging if in waitForAppCompletion mode
-        val loggingInterval = if (waitForAppCompletion) sparkConf.get(REPORT_INTERVAL) else 0
-        val loggingWatch = new LoggingPodStatusWatcher(driverPodCompletedLatch, kubernetesAppId,
-          loggingInterval)
-        Utils.tryWithResource(kubernetesClient
-            .pods()
-            .withName(kubernetesAppId)
-            .watch(loggingWatch)) { _ =>
+      driverServiceManager.start(kubernetesClient, kubernetesAppId, sparkConf)
+      // Begin monitoring the state of the driver pod
+      val loggingInterval = if (waitForAppCompletion) sparkConf.get(REPORT_INTERVAL) else 0
+      val driverPodCompletedLatch = new CountDownLatch(1)
+      val loggingWatch = new LoggingPodStatusWatcher(driverPodCompletedLatch, kubernetesAppId,
+        loggingInterval)
+      Utils.tryWithResource(kubernetesClient
+          .pods()
+          .withName(kubernetesAppId)
+          .watch(loggingWatch)) { _ =>
+        val resourceCleanShutdownHook = ShutdownHookManager.addShutdownHook(() =>
+          kubernetesResourceCleaner.deleteAllRegisteredResourcesFromKubernetes(kubernetesClient))
+        val cleanupServiceManagerHook = ShutdownHookManager.addShutdownHook(
+          ShutdownHookManager.DEFAULT_SHUTDOWN_PRIORITY)(
+          () => driverServiceManager.stop())
+        // Place the error hook at a higher priority in order for the error hook to run before
+        // the stop hook.
+        val serviceManagerErrorHook = ShutdownHookManager.addShutdownHook(
+          ShutdownHookManager.DEFAULT_SHUTDOWN_PRIORITY + 1)(() =>
+          driverServiceManager.handleSubmissionError(
+            new SparkException("Submission shutting down early...")))
+        try {
```

Comment:
@ash211 I swapped the order here of the tryWithResource and try-finally blocks mainly because we want to clean up the secrets and the shutdown hooks after the pod has been launched but also before the application itself completes.